stor/dx.py

```diff
@@ -144,41 +138,273 @@ def download_objects(self):  # may not need it
+                else:
+                    self._clone(dest)
+            else:
+                raise NotFoundError('No data object was found for the given path on DNAnexus')
+        else:
+            super(DXPath, self).copy(dest)
 
     def download_object(self, dest):
-        """Download a single path or object to file."""
-        raise NotImplementedError
+        """Download a single path or object to file.
+
+        Args:
+            dest (Path): The output file
+
+        Raises:
+            ValueError: When source path is not a file
+        """
+        if not self.isfile():
+            raise ValueError('can only call download_object on object path')
+        dxpy.download_dxfile(
+            dxid=self.canonical_resource,
+            filename=dest,
+            project=self.canonical_project
+        )
 
     def download(self, dest):
-        """Download a directory."""
-        raise NotImplementedError
+        """Download a directory.
 
-    def upload(self, source):
-        """Upload a list of files and directories to a directory."""
-        raise NotImplementedError
+        Args:
+            dest (Path): The output directory
+
+        Raises:
+            ValueError: When source or dest path is not a directory
+        """
+        if not self.isdir():
+            raise ValueError('can only call download on folder or project paths')
+        dxpy.download_folder(
+            project=self.canonical_project,
+            destdir=dest,
+            folder=self.resource or '/'
+        )
+
+    def upload(self, to_upload):
+        """Upload a list of files and directories to a directory.
+
+        Note that unlike swift, this is not a batch level operation.
+        If some file errors, the files before will remain uploaded
+
+        Args:
+            to_upload (List): A list of file names, directory names, or
+                OBSUploadObject objects to upload.
+
+        Raises:
+            ValueError: When source path is not a directory
+        """
+        dx_upload_objects = [
+            name for name in to_upload
+            if isinstance(name, OBSUploadObject)
+        ]
+        all_files_to_upload = utils.walk_files_and_dirs([
+            name for name in to_upload
+            if not isinstance(name, OBSUploadObject)
+        ])
+        dx_upload_objects.extend([
+            OBSUploadObject(f,
+                            object_name=(self.resource or Path(''))
+                            / utils.file_name_to_object_name(f))
+            for f in all_files_to_upload
+        ])
+        for upload_obj in dx_upload_objects:
+            upload_obj.object_name = Path(upload_obj.object_name)
+            upload_obj.source = Path(upload_obj.source)
+            dest_file = Path('{drive}{project}:{path}'.format(
+                drive=self.drive, project=self.canonical_project,
+                path=upload_obj.object_name))
+
+            dest_exists = dest_file.isdir()
```

Comment:
can we batch these checks so if you have 100 objects that are uploaded to the same directory you only do one check?

Reply:
Copytree from posix to dx paths is not an atomic operation. It will merge the source and destination folders if they match, like with other obs. This is done to not complicate code and not have special checks for dx everywhere in utils.copy/copytree etc. Hence each destination file/(empty)directory has to be checked if it exists. Can not do a batch check.